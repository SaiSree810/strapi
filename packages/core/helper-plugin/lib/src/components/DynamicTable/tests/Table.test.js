/* eslint-disable react/prop-types */
import {
  BaseCheckbox,
  Tbody,
  Td,
  ThemeProvider,
  Tr,
  Typography,
  lightTheme,
} from '@strapi/design-system';
import { render, fireEvent, waitFor, screen } from '@testing-library/react';
import * as React from 'react';
import { IntlProvider } from 'react-intl';
import { MemoryRouter } from 'react-router-dom';
import TrackingProvider from '../../../providers/TrackingProvider';
import Table from '../index';

const TableRows = ({
  entriesToDelete,
  onSelectRow,
  rows,
  headers,
  withMainAction,
  allowRangeSelection = true,
}) => {
  return (
    <Tbody>
      {rows.map((data, index) => {
        const isChecked = entriesToDelete.findIndex((id) => id === data.id) !== -1;

        const handleRowSelection = (event) => {
          onSelectRow({
            name: data.id,
            value: !isChecked,
            index: allowRangeSelection ? index : undefined,
            isShiftKeyHeld: allowRangeSelection ? event.nativeEvent.shiftKey : undefined,
          });
        };

        return (
          <Tr key={data.id}>
            {withMainAction && (
              <Td>
                <BaseCheckbox
                  aria-label={`Select ${data.firstname} ${data.lastname}`}
                  className={isChecked ? 'selected' : ''}
                  checked={isChecked}
                  onChange={handleRowSelection}
                />
              </Td>
            )}

            {headers.map(({ key, name }) => {
              return (
                <Td key={key}>
                  <Typography textColor="neutral800">{data[name] || '-'}</Typography>
                </Td>
              );
            })}
          </Tr>
        );
      })}
    </Tbody>
  );
};

const renderTable = (headers, rows, allowRangeSelection = true, withMainAction = true) =>
  render(
    <MemoryRouter>
      <TrackingProvider value={{ uuid: '2479d6d7-2497-478d-8a34-a9e8ce45f8a7' }}>
        <ThemeProvider theme={lightTheme}>
          <IntlProvider locale="en" messages={{}} textComponent="span">
            <Table
              headers={headers}
              contentType="users"
              rows={rows}
              withMainAction={withMainAction}
            >
              <TableRows headers={headers} rows={rows} allowRangeSelection={allowRangeSelection} />
            </Table>
          </IntlProvider>
        </ThemeProvider>
      </TrackingProvider>
    </MemoryRouter>
  );

describe('DynamicTable', () => {
  const headers = [
    {
      name: 'firstname',
      metadatas: { label: 'Firstname', sortable: false },
      key: '__firstname_key__',
    },
    {
      name: 'lastname',
      metadatas: { label: 'Lastname', sortable: false },
      key: '__lastname_key__',
    },
    {
      name: 'email',
      metadatas: { label: 'Email', sortable: false },
      key: '__email_key__',
    },
  ];

  const rows = [
    { id: 1, firstname: 'soup', lastname: 'soup', email: 'soup@strapi.io' },
    { id: 2, firstname: 'm', lastname: 'frachet', email: 'm@strapi.io' },
    { id: 3, firstname: 'hicham', lastname: 'ELBSI', email: 'helbsi@strapi.io' },
    { id: 4, firstname: 'john', lastname: 'doe', email: 'doe@strapi.io' },
    { id: 5, firstname: 'mary', lastname: 'jane', email: 'jane@strapi.io' },
  ];

  it('Performs single row item selection/deselection', async () => {
    const { container } = renderTable(headers, rows);

    expect(container).toMatchSnapshot();
    const checkboxes = screen.getAllByRole('checkbox');
    expect(checkboxes).toHaveLength(6);

    checkboxes.map((box) => expect(box).not.toHaveClass('selected'));

    fireEvent.click(checkboxes[1]);
    fireEvent.click(checkboxes[2]);

    await waitFor(() => {
      expect(checkboxes[1]).toHaveClass('selected');
      expect(checkboxes[2]).toHaveClass('selected');
    });

    fireEvent.click(checkboxes[1]);
    fireEvent.click(checkboxes[2]);

    await waitFor(() => {
      expect(checkboxes[1]).not.toHaveClass('selected');
      expect(checkboxes[2]).not.toHaveClass('selected');
    });
  });

  it('Performs range selection/deselection', async () => {
    renderTable(headers, rows);

    const checkboxes = screen.getAllByRole('checkbox');
    expect(checkboxes).toHaveLength(6);

    checkboxes.forEach((box) => expect(box).not.toHaveClass('selected'));

    fireEvent.click(checkboxes[1]);
    fireEvent.click(checkboxes[5], { shiftKey: true });

    await waitFor(() => {
      checkboxes.slice(1, 6).forEach((box) => expect(box).toHaveClass('selected'));
    });

    fireEvent.click(checkboxes[1], { shiftKey: true });
    await waitFor(() => {
      checkboxes.slice(2, 6).forEach((box) => expect(box).not.toHaveClass('selected'));
      expect(checkboxes[1]).toHaveClass('selected');
    });
  });

  it('Perform selection/deselection without range support', async () => {
    renderTable(headers, rows, false);

    const checkboxes = screen.getAllByRole('checkbox');
    expect(checkboxes).toHaveLength(6);

    checkboxes.forEach((box) => expect(box).not.toHaveClass('selected'));

    fireEvent.click(checkboxes[1]);
    fireEvent.click(checkboxes[2]);

    await waitFor(() => {
      expect(checkboxes[1]).toHaveClass('selected');
      expect(checkboxes[2]).toHaveClass('selected');
    });

    fireEvent.click(checkboxes[1]);
    fireEvent.click(checkboxes[2]);

    await waitFor(() => {
      expect(checkboxes[1]).not.toHaveClass('selected');
      expect(checkboxes[2]).not.toHaveClass('selected');
    });
  });
});
