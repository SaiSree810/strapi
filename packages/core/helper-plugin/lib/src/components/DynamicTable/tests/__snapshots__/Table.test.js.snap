// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DynamicTable Performs single row item selection/deselection 1`] = `
.c0 {
  background: #ffffff;
  border-radius: 4px;
  box-shadow: 0px 1px 4px rgba(33,33,52,0.1);
}

.c2 {
  position: relative;
}

.c4 {
  padding-right: 24px;
  padding-left: 24px;
}

.c10 {
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
}

.c13 {
  font-weight: 600;
  font-size: 0.6875rem;
  line-height: 1.45;
  text-transform: uppercase;
  color: #666687;
}

.c15 {
  font-size: 0.875rem;
  line-height: 1.43;
  color: #32324d;
}

.c11 {
  height: 18px;
  min-width: 18px;
  margin: 0;
  border-radius: 4px;
  border: 1px solid #c0c0cf;
  -webkit-appearance: none;
  background-color: #ffffff;
  cursor: pointer;
}

.c11:checked {
  background-color: #4945ff;
  border: 1px solid #4945ff;
}

.c11:checked:after {
  content: '';
  display: block;
  position: relative;
  background: url(data:image/svg+xml;base64,PHN2ZyB3aWR0aD0iMTAiIGhlaWdodD0iOCIgdmlld0JveD0iMCAwIDEwIDgiIGZpbGw9Im5vbmUiIHhtbG5zPSJodHRwOi8vd3d3LnczLm9yZy8yMDAwL3N2ZyI+CiAgPHBhdGgKICAgIGQ9Ik04LjU1MzIzIDAuMzk2OTczQzguNjMxMzUgMC4zMTYzNTUgOC43NjA1MSAwLjMxNTgxMSA4LjgzOTMxIDAuMzk1NzY4TDkuODYyNTYgMS40MzQwN0M5LjkzODkzIDEuNTExNTcgOS45MzkzNSAxLjYzNTkgOS44NjM0OSAxLjcxMzlMNC4wNjQwMSA3LjY3NzI0QzMuOTg1OSA3Ljc1NzU1IDMuODU3MDcgNy43NTgwNSAzLjc3ODM0IDcuNjc4MzRMMC4xMzg2NiAzLjk5MzMzQzAuMDYxNzc5OCAzLjkxNTQ5IDAuMDYxNzEwMiAzLjc5MDMyIDAuMTM4NTA0IDMuNzEyNEwxLjE2MjEzIDIuNjczNzJDMS4yNDAzOCAyLjU5NDMyIDEuMzY4NDMgMi41OTQyMiAxLjQ0NjggMi42NzM0OEwzLjkyMTc0IDUuMTc2NDdMOC41NTMyMyAwLjM5Njk3M1oiCiAgICBmaWxsPSJ3aGl0ZSIKICAvPgo8L3N2Zz4=) no-repeat no-repeat center center;
  width: 10px;
  height: 10px;
  left: 50%;
  top: 50%;
  -webkit-transform: translateX(-50%) translateY(-50%);
  -ms-transform: translateX(-50%) translateY(-50%);
  transform: translateX(-50%) translateY(-50%);
}

.c11:checked:disabled:after {
  background: url(data:image/svg+xml;base64,PHN2ZyB3aWR0aD0iMTAiIGhlaWdodD0iOCIgdmlld0JveD0iMCAwIDEwIDgiIGZpbGw9Im5vbmUiIHhtbG5zPSJodHRwOi8vd3d3LnczLm9yZy8yMDAwL3N2ZyI+CiAgPHBhdGgKICAgIGQ9Ik04LjU1MzIzIDAuMzk2OTczQzguNjMxMzUgMC4zMTYzNTUgOC43NjA1MSAwLjMxNTgxMSA4LjgzOTMxIDAuMzk1NzY4TDkuODYyNTYgMS40MzQwN0M5LjkzODkzIDEuNTExNTcgOS45MzkzNSAxLjYzNTkgOS44NjM0OSAxLjcxMzlMNC4wNjQwMSA3LjY3NzI0QzMuOTg1OSA3Ljc1NzU1IDMuODU3MDcgNy43NTgwNSAzLjc3ODM0IDcuNjc4MzRMMC4xMzg2NiAzLjk5MzMzQzAuMDYxNzc5OCAzLjkxNTQ5IDAuMDYxNzEwMiAzLjc5MDMyIDAuMTM4NTA0IDMuNzEyNEwxLjE2MjEzIDIuNjczNzJDMS4yNDAzOCAyLjU5NDMyIDEuMzY4NDMgMi41OTQyMiAxLjQ0NjggMi42NzM0OEwzLjkyMTc0IDUuMTc2NDdMOC41NTMyMyAwLjM5Njk3M1oiCiAgICBmaWxsPSIjOEU4RUE5IgogIC8+Cjwvc3ZnPg==) no-repeat no-repeat center center;
}

.c11:disabled {
  background-color: #dcdce4;
  border: 1px solid #c0c0cf;
}

.c11:indeterminate {
  background-color: #4945ff;
  border: 1px solid #4945ff;
}

.c11:indeterminate:after {
  content: '';
  display: block;
  position: relative;
  color: white;
  height: 2px;
  width: 10px;
  background-color: #ffffff;
  left: 50%;
  top: 50%;
  -webkit-transform: translateX(-50%) translateY(-50%);
  -ms-transform: translateX(-50%) translateY(-50%);
  transform: translateX(-50%) translateY(-50%);
}

.c11:indeterminate:disabled {
  background-color: #dcdce4;
  border: 1px solid #c0c0cf;
}

.c11:indeterminate:disabled:after {
  background-color: #8e8ea9;
}

.c16 {
  border: 0;
  -webkit-clip: rect(0 0 0 0);
  clip: rect(0 0 0 0);
  height: 1px;
  margin: -1px;
  overflow: hidden;
  padding: 0;
  position: absolute;
  width: 1px;
}

.c1 {
  overflow: hidden;
  border: 1px solid #eaeaef;
}

.c6 {
  width: 100%;
  white-space: nowrap;
}

.c3:before {
  background: linear-gradient(90deg,#c0c0cf 0%,rgba(0,0,0,0) 100%);
  opacity: 0.2;
  position: absolute;
  height: 100%;
  box-shadow: 0px 1px 4px rgba(33,33,52,0.1);
  width: 8px;
  left: 0;
}

.c3:after {
  background: linear-gradient(270deg,#c0c0cf 0%,rgba(0,0,0,0) 100%);
  opacity: 0.2;
  position: absolute;
  height: 100%;
  box-shadow: 0px 1px 4px rgba(33,33,52,0.1);
  width: 8px;
  right: 0;
  top: 0;
}

.c5 {
  overflow-x: auto;
}

.c14 tr:last-of-type {
  border-bottom: none;
}

.c7 {
  border-bottom: 1px solid #eaeaef;
}

.c8 {
  border-bottom: 1px solid #eaeaef;
}

.c8 td,
.c8 th {
  padding: 16px;
}

.c8 td:first-of-type,
.c8 th:first-of-type {
  padding: 0 4px;
}

.c8 th {
  padding-top: 0;
  padding-bottom: 0;
  height: 3.5rem;
}

.c9 {
  vertical-align: middle;
  text-align: left;
  color: #666687;
  outline-offset: -4px;
}

.c9 input {
  vertical-align: sub;
}

.c12 svg {
  height: 0.25rem;
}

<div>
  <div
    class="c0 c1"
  >
    <div
      class="c2 c3"
    >
      <div
        class="c4 c5"
      >
        <table
          aria-colcount="4"
          aria-rowcount="6"
          class="c6"
          role="grid"
        >
          <thead
            class="c7"
          >
            <tr
              aria-rowindex="1"
              class="c8"
            >
              <th
                aria-colindex="1"
                class="c9"
                role="gridcell"
              >
                <div
                  class="c10"
                >
                  <div
                    class=""
                  >
                    <input
                      aria-label="Select all entries"
                      class="c11"
                      tabindex="0"
                      type="checkbox"
                    />
                  </div>
                  <span
                    class="c12"
                  />
                </div>
              </th>
              <th
                aria-colindex="2"
                class="c9"
                role="gridcell"
              >
                <div
                  class="c10"
                >
                  <span>
                    <span
                      aria-labelledby="0"
                      class="c13"
                      label="Firstname"
                      tabindex="-1"
                    >
                      Firstname
                    </span>
                  </span>
                  <span
                    class="c12"
                  />
                </div>
              </th>
              <th
                aria-colindex="3"
                class="c9"
                role="gridcell"
              >
                <div
                  class="c10"
                >
                  <span>
                    <span
                      aria-labelledby="1"
                      class="c13"
                      label="Lastname"
                      tabindex="-1"
                    >
                      Lastname
                    </span>
                  </span>
                  <span
                    class="c12"
                  />
                </div>
              </th>
              <th
                aria-colindex="4"
                class="c9"
                role="gridcell"
              >
                <div
                  class="c10"
                >
                  <span>
                    <span
                      aria-labelledby="2"
                      class="c13"
                      label="Email"
                      tabindex="-1"
                    >
                      Email
                    </span>
                  </span>
                  <span
                    class="c12"
                  />
                </div>
              </th>
            </tr>
          </thead>
          <tbody
            class="c14"
          >
            <tr
              aria-rowindex="2"
              class="c8"
            >
              <td
                aria-colindex="1"
                class="c9"
                role="gridcell"
              >
                <div
                  class=""
                >
                  <input
                    aria-label="Select soup soup"
                    class="c11"
                    tabindex="-1"
                    type="checkbox"
                  />
                </div>
              </td>
              <td
                aria-colindex="2"
                class="c9"
                role="gridcell"
                tabindex="-1"
              >
                <span
                  class="c15"
                >
                  soup
                </span>
              </td>
              <td
                aria-colindex="3"
                class="c9"
                role="gridcell"
                tabindex="-1"
              >
                <span
                  class="c15"
                >
                  soup
                </span>
              </td>
              <td
                aria-colindex="4"
                class="c9"
                role="gridcell"
                tabindex="-1"
              >
                <span
                  class="c15"
                >
                  soup@strapi.io
                </span>
              </td>
            </tr>
            <tr
              aria-rowindex="3"
              class="c8"
            >
              <td
                aria-colindex="1"
                class="c9"
                role="gridcell"
              >
                <div
                  class=""
                >
                  <input
                    aria-label="Select m frachet"
                    class="c11"
                    tabindex="-1"
                    type="checkbox"
                  />
                </div>
              </td>
              <td
                aria-colindex="2"
                class="c9"
                role="gridcell"
                tabindex="-1"
              >
                <span
                  class="c15"
                >
                  m
                </span>
              </td>
              <td
                aria-colindex="3"
                class="c9"
                role="gridcell"
                tabindex="-1"
              >
                <span
                  class="c15"
                >
                  frachet
                </span>
              </td>
              <td
                aria-colindex="4"
                class="c9"
                role="gridcell"
                tabindex="-1"
              >
                <span
                  class="c15"
                >
                  m@strapi.io
                </span>
              </td>
            </tr>
            <tr
              aria-rowindex="4"
              class="c8"
            >
              <td
                aria-colindex="1"
                class="c9"
                role="gridcell"
              >
                <div
                  class=""
                >
                  <input
                    aria-label="Select hicham ELBSI"
                    class="c11"
                    tabindex="-1"
                    type="checkbox"
                  />
                </div>
              </td>
              <td
                aria-colindex="2"
                class="c9"
                role="gridcell"
                tabindex="-1"
              >
                <span
                  class="c15"
                >
                  hicham
                </span>
              </td>
              <td
                aria-colindex="3"
                class="c9"
                role="gridcell"
                tabindex="-1"
              >
                <span
                  class="c15"
                >
                  ELBSI
                </span>
              </td>
              <td
                aria-colindex="4"
                class="c9"
                role="gridcell"
                tabindex="-1"
              >
                <span
                  class="c15"
                >
                  helbsi@strapi.io
                </span>
              </td>
            </tr>
            <tr
              aria-rowindex="5"
              class="c8"
            >
              <td
                aria-colindex="1"
                class="c9"
                role="gridcell"
              >
                <div
                  class=""
                >
                  <input
                    aria-label="Select john doe"
                    class="c11"
                    tabindex="-1"
                    type="checkbox"
                  />
                </div>
              </td>
              <td
                aria-colindex="2"
                class="c9"
                role="gridcell"
                tabindex="-1"
              >
                <span
                  class="c15"
                >
                  john
                </span>
              </td>
              <td
                aria-colindex="3"
                class="c9"
                role="gridcell"
                tabindex="-1"
              >
                <span
                  class="c15"
                >
                  doe
                </span>
              </td>
              <td
                aria-colindex="4"
                class="c9"
                role="gridcell"
                tabindex="-1"
              >
                <span
                  class="c15"
                >
                  doe@strapi.io
                </span>
              </td>
            </tr>
            <tr
              aria-rowindex="6"
              class="c8"
            >
              <td
                aria-colindex="1"
                class="c9"
                role="gridcell"
              >
                <div
                  class=""
                >
                  <input
                    aria-label="Select mary jane"
                    class="c11"
                    tabindex="-1"
                    type="checkbox"
                  />
                </div>
              </td>
              <td
                aria-colindex="2"
                class="c9"
                role="gridcell"
                tabindex="-1"
              >
                <span
                  class="c15"
                >
                  mary
                </span>
              </td>
              <td
                aria-colindex="3"
                class="c9"
                role="gridcell"
                tabindex="-1"
              >
                <span
                  class="c15"
                >
                  jane
                </span>
              </td>
              <td
                aria-colindex="4"
                class="c9"
                role="gridcell"
                tabindex="-1"
              >
                <span
                  class="c15"
                >
                  jane@strapi.io
                </span>
              </td>
            </tr>
          </tbody>
        </table>
      </div>
    </div>
  </div>
  <div
    class="c16"
  >
    <p
      aria-live="polite"
      aria-relevant="all"
      id="live-region-log"
      role="log"
    />
    <p
      aria-live="polite"
      aria-relevant="all"
      id="live-region-status"
      role="status"
    />
    <p
      aria-live="assertive"
      aria-relevant="all"
      id="live-region-alert"
      role="alert"
    />
  </div>
</div>
`;
